bootstrapvz/plugins/admin_user/tasks.py

```diff
@@ -41,14 +44,89 @@ def run(cls, info):
+		log = logging.getLogger(__name__)
+
+		log.debug('Setting the password for the admin user.')
+		log_check_call(
+			['chroot', info.root, 'chpasswd'],
+			info.manifest.plugins['admin_user']['username'] +
+			':' + info.manifest.plugins['admin_user']['password']
+		)
+		return
+
+
+class AdminUserCredentialsPublicKey(Task):
+	description = 'Set up access credentials for the admin user with a given public key'
+	phase = phases.system_modification
+	predecessors = [AddEC2InitScripts, CreateAdminUser]
+	successors = [InstallInitScripts]
+
+	@classmethod
+	def run(cls, info):
+		from bootstrapvz.common.tools import log_check_call
+
+		log = logging.getLogger(__name__)
+
+		import stat
+		from shutil import copy
+		full_path = info.manifest.plugins['admin_user']['pubkey']
+		log.debug('Copying public key from {path}'.format(path=full_path))
+
+		if 'ec2-get-credentials' in info.initd['install']:
+			log.warn(
+				'You are using a static public key for the admin account.'
+				' This will conflict with the ec2 public key injection mechanisn.'
+				' The ec2-get-credentials startup script has therefore been disabled.')
+			del info.initd['install']['ec2-get-credentials']
+
+		username = info.manifest.plugins['admin_user']['username']
+
+		ssh_file = os.path.join('/home/', username, '.ssh/authorized_keys')
+		rel_ssh_file = os.path.realpath(info.root + '/%s' % ssh_file)
+
+		ssh_dir = os.path.dirname(ssh_file)
+		rel_ssh_dir = os.path.realpath(info.root + '/%s' % ssh_dir)
+		if not os.path.exists(rel_ssh_dir):
+			log.debug('Creating %s.' % rel_ssh_dir)
+			os.mkdir(rel_ssh_dir)
+
+		log.debug('setting %s mode 700' % rel_ssh_dir)
+		mode = (stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+		os.chmod(rel_ssh_dir, mode)
+
+		copy(full_path, rel_ssh_file)
+
+		mode = (stat.S_IRUSR | stat.S_IWUSR)
+		os.chmod(rel_ssh_file, mode)
+
+		log_check_call(['chroot', info.root, 'chown', '-R', username, ssh_dir])
+		return
+
+
+class AdminUserCredentialsEC2(Task):
+	description = 'Set up access credentials for the admin user using the EC2 credentials'
 	phase = phases.system_modification
-	predecessors = [InstallInitScripts]
+	predecessors = [InstallInitScripts, CreateAdminUser]
 
 	@classmethod
 	def run(cls, info):
+		from bootstrapvz.common.exceptions import TaskError
 		from bootstrapvz.common.tools import sed_i
+		log = logging.getLogger(__name__)
+
 		getcreds_path = os.path.join(info.root, 'etc/init.d/ec2-get-credentials')
-		username = info.manifest.plugins['admin_user']['username']
-		sed_i(getcreds_path, 'username=\'root\'', 'username=\'{username}\''.format(username=username))
+		if os.path.exists(getcreds_path):
+			log.debug('Updating EC2 get credentials script.')
+			username = info.manifest.plugins['admin_user']['username']
+			sed_i(getcreds_path, "username='root'",
+			      "username='{username}'".format(username=username))
+		else:
+			raise TaskError('Could not find EC2 get credentials script.')
```

Comment:
No need for this really. `sed_i` should fail if the path does not exist, which would be a coding errors since it should exist at this point - so just letting `sed_i` fail is fine :-)